### ISSUE TITLE
Suggestion: Skip support (SkippableFact)

### ISSUE THREAD

**Issue by bent-rasmussen** [AArnott/Xunit.StaFact#102]
Trying to do UI testing with `StaFact` but then I would also like to be able to skip tests. However, adding both a `SkippableFact` and a `StaFact` attribute causes xunit to fail on not supporting multiple `Fact`-derived attributes. If trying to call `Skip.If(true)` anyway, then the `Xunit.SkipException` is thrown and unhandled, causing the test case to fail.

**Comment by AArnott** [comment]
It's fair feedback. It personally blows my mind that Xunit doesn't have its own skippable indicator. And crossing between two xunit fact types gets muddy.  I think the next major version of xunit is supposed to address this gap anyway.
But in the meantime, hey, I could see adding support to report SkipException as skips in StaFact as well.
Would you be interested in submitting the PR with the change yourself?

**Comment by bent-rasmussen** [comment]
Yes, it's quite surprising that proper skip support is not built into xunit. I did not know that it is planned for v3. I'd be inclined to wait but I could take a look and see how involved the task is for someone who hasn't worked with xunit internals before. On the surface it would seem appropriate to take a dependency on SkippableFact and make the other attributes inherit from Skippable*.

**Comment by AArnott** [comment]
> it would seem appropriate to take a dependency on SkippableFact and make the other attributes inherit from Skippable*.

Ah, if only it were that easy. But no, I'm pretty sure that won't work.

Instead, I suggest you look at _how_ `SkippableFact` does it and apply the same pattern to the StaFact attributes. You'll first notice that it isn't the attribute itself at all. It's the test executor that the attribute activates. Each library has their own executor, which is why inheritance alone won't work.
I also don't think the StaFact package should depend on the Skippable package. So instead, I think the StaFact system should simply do a name check on the exception that's thrown (so there is no type reference to the Skippable assembly) and if it matches, it can report a skipped test in the same way that the Skippable package does.

**Comment by bent-rasmussen** [comment]
_So instead, I think the StaFact system should simply do a name check on the exception that's thrown (so there is no type reference to the Skippable assembly) and if it matches, it can report a skipped test in the same way that the Skippable package does._

That was my first thought as well. I just figured there was much duplication of work otherwise - MessageBus, etc., assuming partial reuse would be possible. But I take your point and the SkippableFact package seems pretty small compared to the StaFact package.

**Comment by AArnott** [comment]
xUnit v3 has built-in support for skipping. Alas, it won't (all) apply automatically to xunit.stafact (even in its v3 supporting version) because some of the support is in the attributes themselves. We may also need to update our test runners. So let's keep this active.

@reduckted are you interested in picking this one up?

**Comment by reduckted** [comment]
Yeah, I'll take a look and see what's needed. 👍

**Comment by reduckted** [comment]
This appears to be fully supported with #118 (with the caveat that exception-based skipping will arrive in xUnit 3 v2.0 and will require some minor changes - https://github.com/xunit/xunit/issues/3101).

With these tests:
```cs
[StaFact(Skip = "testing skipping.")]
public void CanSkip()
{
    Assert.Fail("Should be skipped.");
}

[StaFact(Skip = "Conditionally skipping", SkipWhen = nameof(SkipIfEvenMinute))]
public void CanSkipWhenEvenMinute()
{
    Assert.Equal(1, DateTime.Now.Minute % 2);
}

[StaFact(Skip = "Conditionally skipping", SkipUnless = nameof(SkipIfEvenMinute))]
public void CanSkipUnlessEvenMinute()
{
    Assert.Equal(0, DateTime.Now.Minute % 2);
}

[StaFact]
public void CanSkipInTest()
{
    Assert.Skip("even minute");
}

[StaFact]
public void CanSkipWhenEvenMinuteInTest()
{
    Assert.SkipWhen(SkipIfEvenMinute, "even minute");
}

[StaFact]
public void CanSkipUnlessEvenMinuteInTest()
{
    Assert.SkipUnless(SkipIfEvenMinute, "odd minute");
}

public static bool SkipIfEvenMinute => (DateTime.Now.Minute % 2) == 0;
```
This is the result when running the tests on an even minute:
![Test results from even minute](https://github.com/user-attachments/assets/3038cd75-fde9-4dfc-a4fa-8ef6baac4acf)

And this is the result when running the tests on an odd minute:

![Test results from odd minute](https://github.com/user-attachments/assets/8a38c339-9743-4073-a08d-93c2c638bfb6)

**Comment by reduckted** [comment]
> This appears to be fully supported with [#118](https://github.com/AArnott/Xunit.StaFact/pull/118) (with the caveat that exception-based skipping will arrive in xUnit 3 v2.0 and will require some minor changes - [xunit/xunit#3101](https://github.com/xunit/xunit/issues/3101)).

xUnit 3 v2.0 has been released. I don't remember what "will require some minor changes", but I'll take another look and see what needs to change.
